### ISSUE TITLE
gtk::init() is missing a test for acquiring context

### ISSUE THREAD

**Issue by ArekPiekarz** [gtk-rs/gtk3-rs#616]
This is a followup of conversation from https://github.com/gtk-rs/gtk3-rs/pull/614.

`gtk::init()` in gtk3-rs is missing a unit test for acquiring a context, which is present in gtk4-rs - https://github.com/gtk-rs/gtk4-rs/blob/master/gtk4/src/rt.rs#L130. It was not included because of issues with synchronization when running `cargo test`, which by default launches tests in multiple threads. gtk4-rs has [TEST_THREAD_WORKER](https://github.com/gtk-rs/gtk4-rs/blob/master/gtk4/src/lib.rs#L63) for solving that problem.

I see two options for now to proceed with this topic:
1. Duplicate TEST_THREAD_WORKER in gtk3-rs, or
2. Move TEST_THREAD_WORKER to a shared repository (gtk-rs-core), make that variable public and use it in gtk3-rs and gtk4-rs. This has an added benefit that users of these libraries would also have access to that synchronization tool in their tests.

After one of the above options is done, the unit test could be ported to gtk3-rs. What do you think about it?

**Comment by GuillaumeGomez** [comment]
If you go for the second proposition, we have https://github.com/gtk-rs/checker which is used for "common" tests. So as you prefer.

**Comment by ArekPiekarz** [comment]
Can you elaborate how the `checker` would help with the second option? The closest thing in there to our problem would be https://github.com/gtk-rs/checker/blob/master/check_init_asserts, but it only checks if we have some kind of assertion for `init()`, it doesn't provide synchronization for tests. Or do you want to put TEST_THREAD_WORKER in the checker repository?

**Comment by GuillaumeGomez** [comment]
You were talking about using a shared repository, I simply gave you one we already have. After it's up to you to check how you want to do it in it. ;)

**Comment by ArekPiekarz** [comment]
I looked at it deeper and found that in order for a crate to be published on crates.io, all its dependencies also have to be crates there. Currently I do not see any existing shared dependency for gtk and gtk4 that would be a good place to store TEST_THREAD_WORKER. I guess your organization would need to create a new crate just for it, which seems a little overkill. 
As for the checker, I found its usage only in CI, not in the library tests.
Please correct me if I got anything wrong.

To conclude, for now I will try to go with the first option.

**Comment by GuillaumeGomez** [comment]
No you're absolutely right, we use this crate only in the CI (it runs some common checks). And as you prefer. :)

**Comment by ArekPiekarz** [comment]
While I've been porting the test, I noticed something strange - no matter what I've put inside the closure pushed to TEST_THREAD_WORKER, the test always passed. Then at the end of the test, outside the closure I added:
`assert_eq!(TEST_THREAD_WORKER.unprocessed(), 0);`
and it failed, showing 2 unprocessed tasks.

Only after I replaced the above with:
`while TEST_THREAD_WORKER.unprocessed() > 0 {}`
the test passed or fail, depending on the content of the closure.
It looks to me as if the current tests in gtk4-rs do not wait for the pool to finish its tasks. If they work now, it may be by accident. I'll create a separate issue for that later.
